source/common/network/address_impl.cc

```diff
@@ -352,11 +367,25 @@ absl::Status Ipv6Instance::validateProtocolSupported() {
   return absl::FailedPreconditionError("IPv6 addresses are not supported on this machine");
 }
 
-void Ipv6Instance::initHelper(const sockaddr_in6& address, bool v6only) {
+void Ipv6Instance::initHelper(const sockaddr_in6& address, bool v6only,
+                              bool generate_address_string) {
   ip_.ipv6_.address_ = address;
-  ip_.friendly_address_ = ip_.ipv6_.makeFriendlyAddress();
   ip_.ipv6_.v6only_ = v6only;
-  friendly_name_ = fmt::format("[{}]:{}", ip_.friendly_address_, ip_.port());
+  if (generate_address_string) {
+    ip_.address_string_ = ip_.ipv6_.makeFriendlyAddress();
+    setAddressPortString(generateAddressPortString());
+  }
+}
+
+const std::string& Ipv6Instance::IpHelper::addressAsString() const {
+  if (address_string_.empty()) {
+    address_string_ = ipv6_.makeFriendlyAddress();
```

Comment:
I don't think this is thread-safe. Addresses are often constructed and then passed around as `Network::Address::InstanceConstSharedPtr`, and it is assumed that all the `const` methods on them are thread-safe.

I wonder if we could get away with not storing/caching the string representation, and generating a new string everytime it's called. 

I'm not sure what use cases that may degrade performance on. 

Another option is to add a mutex here.

Reply:
I think this is mostly used in logging. It'd hard to tell if not caching the string form will cause performance degradation given access log is an extension.

I can use std::call_once here if it is any better than mutex?

Reply:
If we only use this functionality in logging I wonder how frequently we use this method multiple times vs a single time. If it's typically a single usage, then we could just have a method which returns a new string, and doing so on-demand, instead of storing it in a member and returning a ref to the member.

Reply:
Agree that `call_once` is better than a mutex, and sufficient for what we need here.

It would be a much more involved investigation (and change) to stop caching them I think, but might be worthwhile to investigate at some point.

Reply:
Changed it to use call_once.